From 8b91169789ccf8956228cb118112dc4ee694a56d Mon Sep 17 00:00:00 2001
From: Luca Fancellu <luca.fancellu@arm.com>
Date: Tue, 13 Jun 2023 11:31:05 +0100
Subject: [PATCH] net: config: Add VLAN identifier as network initial
 configuration

Add a new Kconfig parameter NET_CONFIG_MY_VLAN_ID as initial network
configuration to enable users to set VLAN identifier at startup.

Add a call in net_config_init_by_iface(...) function to setup the
VLAN identifier in the device if the Kconfig parameter is different
from zero.

Signed-off-by: Luca Fancellu <luca.fancellu@arm.com>
Upstream-Status: Backport [main: 38a43b0eec494a0bce0d94c8c31309c9cfe67984]
---
 subsys/net/lib/config/Kconfig |  7 +++++++
 subsys/net/lib/config/init.c  | 14 ++++++++++++++
 2 files changed, 21 insertions(+)

diff --git a/subsys/net/lib/config/Kconfig b/subsys/net/lib/config/Kconfig
index 528bbb6986f7..0f1e3c50544e 100644
--- a/subsys/net/lib/config/Kconfig
+++ b/subsys/net/lib/config/Kconfig
@@ -71,6 +71,13 @@ menuconfig NET_CONFIG_SETTINGS
 
 if NET_CONFIG_SETTINGS
 
+config NET_CONFIG_MY_VLAN_ID
+	int "My VLAN identifier"
+	default 0
+	depends on NET_VLAN
+	help
+	  Use 0 here if uncertain.
+
 if NET_IPV6
 
 config NET_CONFIG_MY_IPV6_ADDR
diff --git a/subsys/net/lib/config/init.c b/subsys/net/lib/config/init.c
index 219852df81eb..edb34a08bba5 100644
--- a/subsys/net/lib/config/init.c
+++ b/subsys/net/lib/config/init.c
@@ -16,6 +16,7 @@ LOG_MODULE_REGISTER(net_config, CONFIG_NET_CONFIG_LOG_LEVEL);
 #include <stdlib.h>
 
 #include <zephyr/logging/log_backend.h>
+#include <zephyr/net/ethernet.h>
 #include <zephyr/net/net_core.h>
 #include <zephyr/net/net_ip.h>
 #include <zephyr/net/net_if.h>
@@ -378,6 +379,19 @@ int net_config_init_by_iface(struct net_if *iface, const char *app_info,
 #endif
 	}
 
+#if defined(CONFIG_NET_VLAN)
+	if (CONFIG_NET_CONFIG_MY_VLAN_ID > 0) {
+		int ret = net_eth_vlan_enable(iface,
+					      CONFIG_NET_CONFIG_MY_VLAN_ID);
+
+		if (ret < 0) {
+			NET_ERR("Network interface %d (%p): cannot set VLAN tag (%d)\n",
+				net_if_get_by_iface(iface), iface, ret);
+			return -ENOEXEC;
+		}
+	}
+#endif
+
 	setup_ipv4(iface);
 	setup_dhcpv4(iface);
 	setup_ipv6(iface, flags);
-- 
2.34.1

